File: app/src/main/java/com/example/cinemates/adapter/SectionRecyclerViewAdapter.java
package com.example.cinemates.adapter;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.example.cinemates.databinding.SectionRowBinding;
import com.example.cinemates.model.Movie;
import com.example.cinemates.util.RecyclerViewEmptySupport;
import com.example.cinemates.views.fragment.HomeFragmentDirections;
import com.example.cinemates.model.Section;

import java.util.ArrayList;
import java.util.ConcurrentModificationException;
import java.util.List;

/**
 * @author Antonio Di Nuzzo
 * Created 15/12/2021 at 16:36
 */
public class SectionRecyclerViewAdapter<T> extends RecyclerView.Adapter<SectionRecyclerViewAdapter.SectionViewHolder> {
    private final List<Section<T>> dataList = new ArrayList<>();
    private final LifecycleOwner mLifecycleOwner;

    public SectionRecyclerViewAdapter(LifecycleOwner lifecycleOwner) {
        mLifecycleOwner = lifecycleOwner;
    }

    @NonNull
    @Override
    public SectionViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        SectionRowBinding sectionRowBinding = SectionRowBinding.inflate(layoutInflater, parent, false);
        return new SectionViewHolder(sectionRowBinding);
    }

    @Override
    public void onBindViewHolder(SectionViewHolder holder, int position) {
        Section<T> section = dataList.get(position);
        holder.mBinding.setSection(section);
        holder.mBinding.executePendingBindings();

        SectionItemsRecyclerViewAdapter<T> sectionItemsRecyclerViewAdapter = new SectionItemsRecyclerViewAdapter<>();
        holder.mBinding.recyclerView.setAdapter(sectionItemsRecyclerViewAdapter);
        holder.mBinding.recyclerView.setEmptyView(holder.mBinding.emptyView.getRoot());
        section.getMutableLiveData().observe(mLifecycleOwner, new Observer<ArrayList<T>>() {
            @Override
            public void onChanged(ArrayList<T> items) {
                sectionItemsRecyclerViewAdapter.addItems((items));

            }
        });


    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    public void addItems(List<Section<T>> dataList) {
        this.dataList.addAll(dataList);
        notifyDataSetChanged();
    }

    public void addItems(Section<T> section) {
        this.dataList.add(section);
        notifyDataSetChanged();
    }

    static class SectionViewHolder extends RecyclerViewEmptySupport.ViewHolder {
        SectionRowBinding mBinding;

        SectionViewHolder(@NonNull SectionRowBinding sectionRowBinding) {
            super(sectionRowBinding.getRoot());
            this.mBinding = sectionRowBinding;

            this.mBinding.actionOpenPage.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {

                  /*  HomeFragmentDirections.ActionHomeFragmentToDetailedViewFragment action =
                            HomeFragmentDirections.actionHomeFragmentToDetailedViewFragment();
                    action.setSection(mBinding.textSectionTitle.getText().toString());
                    Navigation.findNavController(view).navigate(action);*/
                    Toast.makeText(view.getContext(), "On develop", Toast.LENGTH_SHORT).show();

                }
            });


        }


    }
}

File: app/src/main/java/com/example/cinemates/view/ui/adapter/SectionRecyclerViewAdapter.kt
package com.example.cinemates.view.ui.adapter

import android.view.ViewGroup
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.AsyncDifferConfig
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.cinemates.view.ui.adapter.SectionRecyclerViewAdapter.SectionViewHolder
import com.example.cinemates.databinding.ListItemSectionBinding
import com.example.cinemates.model.*
import com.example.cinemates.util.inflater

/**
 * @author Antonio Di Nuzzo
 * Created 15/12/2021 at 16:36
 */
class SectionRecyclerViewAdapter(private val lifecycleOwner: LifecycleOwner) :
    ListAdapter<Section<*>, SectionViewHolder>(asyncDiffConfig) {
    private val dataList: MutableList<Section<*>> = mutableListOf()

    private companion object {
        private const val PERSON = 0
        private const val MOVIE = 1
        private const val TV = 2
        private val diffCallback = object : DiffUtil.ItemCallback<Section<*>>() {
            override fun areItemsTheSame(oldItem: Section<*>, newItem: Section<*>): Boolean =
                oldItem.liveData == newItem.liveData

            override fun areContentsTheSame(oldItem: Section<*>, newItem: Section<*>): Boolean =
                areItemsTheSame(oldItem, newItem)
        }

        private val asyncDiffConfig = AsyncDifferConfig.Builder(diffCallback).build()
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SectionViewHolder {
        return SectionViewHolder(parent inflater ListItemSectionBinding::inflate)
    }

    override fun onBindViewHolder(holder: SectionViewHolder, position: Int) {
        when (holder.itemViewType) {
            MOVIE -> {
                val movieSection = dataList[position] as Section<Movie>
                val sectionItemsMovie = MovieAdapter()
                holder.binding.apply {
                    section = movieSection
                    executePendingBindings()
                    recyclerView.adapter = sectionItemsMovie
                    recyclerView.setEmptyView(holder.binding.emptyView.root)
                }
                movieSection.liveData.observe(lifecycleOwner) { items ->
                    sectionItemsMovie.addItems(
                        items
                    )
                }

            }
            TV -> {
                val tvShowSection = dataList[position] as Section<TvShow>
                val sectionItemsTv = TvShowAdapter()
                holder.binding.apply {
                    section = tvShowSection
                    executePendingBindings()
                    recyclerView.adapter = sectionItemsTv
                    recyclerView.setEmptyView(holder.binding.emptyView.root)
                }
                tvShowSection.liveData.observe(lifecycleOwner) { items ->
                    sectionItemsTv.addItems(
                        items
                    )
                }

            }
            PERSON -> {
                val personSection = dataList[position] as Section<Person>
                val sectionItemsPerson = PersonAdapter()
                holder.binding.apply {
                    section = personSection
                    executePendingBindings()
                    recyclerView.adapter = sectionItemsPerson
                    recyclerView.setEmptyView(holder.binding.emptyView.root)
                }
                personSection.liveData.observe(lifecycleOwner) { items ->
                    sectionItemsPerson.addItems(
                        items
                    )
                }
            }
        }
    }


    override fun getItemCount(): Int {
        return dataList.size
    }

    fun addItems(dataList: List<Section<*>>) {
        this.dataList.clear()
        this.dataList.addAll(dataList)
        notifyDataSetChanged()
    }

    fun addItems(section: Section<*>) {
        dataList.add(section)
        notifyDataSetChanged()
    }

    //Returns the view type of the item at position for the purposes of view recycling.
    override fun getItemViewType(position: Int): Int {
        val value = dataList[position].genericType

        return when (value) {
            Movie::class.java -> MOVIE
            Person::class.java, Cast::class.java -> PERSON
            TvShow::class.java -> TV
            else -> -1
        }
    }

    class SectionViewHolder(val binding: ListItemSectionBinding) :
        RecyclerView.ViewHolder(
            binding.root
        )
}